Add support for request interception patterns

paambaati opened the pull request:
This PR introduces support for request interception patterns, freshly introduced in https://chromium-review.googlesource.com/c/chromium/src/+/602483

I'm not sure which release this commit landed in, as I can't find it yet on https://omahaproxy.appspot.com/

The new test fails for now, but rolling Chromium to the version that includes the commit should make it pass.

paambaati committed:
Add support for request interception patterns.

This feature is introduced in https://chromium-review.googlesource.com/c/chromium/src/+/602483

paambaati committed:
Tests for pattern-based request interception.

paambaati commented:
@googlebot I signed it!

JoelEinbinder commented on test/test.js:
```diff
@@ -810,6 +810,17 @@ describe('Page', function() {
+    it('should block patterns', SX(async function() {
+      await page.setRequestInterceptionEnabled(true, ['*.css']);
+      page.on('request', request => {
+        request.continue();
```
Shouldn't this be request.abort()?

paambaati replied:
D'oh, you're right. Fixed in 72800b7.

paambaati replied:
@JoelEinbinder Also, the test would be more rounded if I have more XHRs in a page and I block specific patterns, while leaving out the rest. Should I create another HTML fixture?

aslushnikov committed:
Fix Request Interception (#565)

It turned out that either Network.requestIntercepted or
Network.requestWillBeSent occasionally report encoded URL.

This patch starts decoding URL's when generating request hash.

Fixes #558.

twalpole commented:
I'm unclear as to what this option actually does.  Does it mean the `request` event is only emitted when a request to a matching URL happens, or does it mean that request interception is only enabled for requests to matching URLS?  If it's the second, is there a way inside the `request` event handler to determine whether or not interception has been enabled or not (whether or not abort/continue need to be called)?

Mohamed3on committed:
doc: Clarify that examples need Node v8 (#559)

The examples use async/await, but Node v6 doesn't support them, so to use these examples you either need Node 8 or a transpiler (like Babel) or to use promises.

aslushnikov commented:
@paambaati thank you for the PR!

Why would you want this? The same behavior is implementable with the request interception API puppeteer has today:

```js
const blockRegexes = [
  /\.jpg$/i,
  /\.jpeg$/i,
  /\.png$/i,
];
page.on('request', request => {
  if (blockRegexes.some(regex => regex.test(request.url))
    request.abort();
  else
    request.continue();
});
```

paambaati committed:
We should abort the request as the test is to block pattern-matched requests.

paambaati commented:
@twalpole The `requestIntercepted` event is fired only for request URLs that match the patterns.

paambaati commented:
@aslushnikov I'd say this is a convenience method. If the protocol comes bundled with it, I'd rather use that instead of the boilerplate from the example.

EDIT: In addition, I believe things should be overall faster as we're not even intercepting them. I got interested after this in the [commit message](https://chromium-review.googlesource.com/c/chromium/src/+/602483) —

> _This should increase the stalled speed a bit._

twalpole commented:
@paambaati  Are you saying there is a new `requestIntercepted` event, because your test only uses the `request` and `requestFailed` events?  

Normally the `request` event is fired  for every request,  and if interception is enabled you need to call `abort` or `continue` on the `request` object.  If interception is not enabled then you can't call `abort` or `continue` on the`request` object.  I'm guessing what this PR actually does is automatically enable request interception based on the patterns (but not limit which requests the request event is fired for). However if there is no way in the `request` event handler to know whether interception is enabled or not, then you have to check the patterns against the url in the handler to determine whether or not `abort`/`continue` need to be called which seems highly inconvenient and defeats the point of this PR.

paambaati commented:
@twalpole `requestIntercepted` is a wire protocol event, I don't think puppeteer exposes it. Sorry for the confusion!

To answer your question, pupeteer's `request` event will be fired for each intercepted request, which includes pattern-matched ones.

> However if there is no way in the request event handler to know whether interception is enabled or not

You're correct, there isn't. See https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#requestabort

paambaati commented:
@JoelEinbinder @aslushnikov Looks like this landed in [`62.0.3197.0`](https://storage.googleapis.com/chromium-find-releases-static/241.html#241c6b4df3f4c0b93d783ed2b5b8b448e4e2d285).

paambaati committed:
Default interception patterns to ['*'].

This is so if patterns are not given, we intercept all requests.

aslushnikov commented:
@paambaati The main use case of pattern-blocking is to block all requests of the same type, e.g. all images or all styles. There's a better way to do it using resource type: check out #590 for details.

Codewise, I don't like the suggested approach because of the following questions:
    - When is it ok to call `request.continue()` and when not?
    - How does URL blocking work with redirects?
    - How does URL blocking work with HTTPS errors?
    - How do I know if the request's URL matches the request blocking pattern?

Unless there are very good reasons for this PR and good answers to the questions, I'd rather not land this.

paambaati commented:
@aslushnikov Fair enough. TBH, I agree with the observation that this doesn't go well with the `request` event and `request.continue()`.
